### PR TITLE
[NR-165205] Add codecov

### PR DESCRIPTION
**Pull request by ywang-nr**
**SQUASH TO MERGE**
1. Added codecov for all modules
2. Create GHA task

### CODE REVIEW

File: agent/build.gradle [hunk 1]
```diff
@@ -84,7 +90,7 @@ dependencies {
     compileOnly fileTree(dir: 'libs', include: '*.jar')
     compileOnly newrelic.deps.ndk
 
-compileOnly "androidx.navigation:navigation-compose:${project.versions.jetpack}"
+    compileOnly "androidx.navigation:navigation-compose:${project.versions.jetpack}"
```
Comment (cthomas-newrelic) [inline agent/build.gradle:93]: Thanks for fixing this!

File: .github/workflows/codecov.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        fetch-depth: 0
+          
+    - name: set up JDK 11
+      uses: actions/setup-java@v2
+      with:
+        java-version: '11'
+        distribution: 'adopt'
+        cache: gradle
+
+    - name: Grant execute permission for gradlew
```
Comment (cthomas-newrelic) [inline .github/workflows/codecov.yml:25]: Use the Gradle build action as is done [here](https://github.com/newrelic/newrelic-android-agent/blob/develop/.github/workflows/functionalTests.yml#L22-L25).

File: .github/workflows/codecov.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      run: chmod +x gradlew
+
+    - name: Run jacoco
+      run: |
+        ./gradlew agent:testReleaseUnitTestCoverage
+        ./gradlew agent-core:jacocoTestReport
+        ./gradlew instrumentation:jacocoTestReport
+        ./gradlew plugin:gradle:jacocoTestReport
+
+    # - name: Generate report
```
Comment (cthomas-newrelic) [inline .github/workflows/codecov.yml:35]: Is the commented code worth keeping? Will it be enabled?

File: jacoco.gradle [hunk 4]
```diff
@@ -0,0 +1,93 @@
+private static boolean isAndroidModule(Project project) {
+    boolean isAndroidLibrary = project.plugins.hasPlugin('com.android.library')
+    boolean isAndroidApp = project.plugins.hasPlugin('com.android.application')
+    return isAndroidLibrary || isAndroidApp
+}
+
+afterEvaluate { project ->
+    def projectName = project.name
+    if (isAndroidModule(project)) setupAndroidReporting()
```
Comment (cthomas-newrelic) [inline jacoco.gradle:9]: Formatting: ```
if {
   // if code
} else {
   // else code
}```

Or let `Shift-Command-L` do it for you.

File: jacoco.gradle [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    jacocoTestReport {
+        dependsOn test
+        reports {
+            xml.required = true
+            html.required = true
+        }
+        afterEvaluate {
+            classDirectories.from = files(classDirectories.files.collect {
+                fileTree(dir: it, exclude: [
+                        // dagger
```
Comment (cthomas-newrelic) [inline jacoco.gradle:23]: What's all this? We don't use Dagger or Hilt in the agent. It looks like it was taken from an example.

File: jacoco.gradle [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    // Grab all build types and product flavors
+    def buildTypes = android.buildTypes.collect { type ->
+        type.name
+    }
+    def productFlavors = android.productFlavors.collect { flavor ->
+        flavor.name
+    }
+    // When no product flavors defined, use empty
+    if (!productFlavors) productFlavors.add('')
```
Comment (cthomas-newrelic) [inline jacoco.gradle:56]: Formatting: ``` 
if() {
   // if code
 }```

Comment (cthomas-newrelic) [reply jacoco.gradle:56]: Sorry , GitHub is not properly formatting the comments. Should like like:

if() {
   // if code
}

File: jacoco.gradle [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        buildTypes.each { buildTypeName ->
+            def sourceName, sourcePath
+            if (!productFlavorName) {
+                sourceName = sourcePath = "${buildTypeName}"
+            } else {
+                sourceName = "${productFlavorName}${buildTypeName.capitalize()}"
+                sourcePath = "${productFlavorName}/${buildTypeName}"
+            }
+            def testTaskName = "test${sourceName.capitalize()}UnitTest"
+            System.out.println("Task -> $testTaskName")
```
Comment (cthomas-newrelic) [inline jacoco.gradle:67]: Use the logger provided by Gradle.

File: jacoco.gradle [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+            // Create coverage task of form 'testFlavorTypeCoverage' depending on 'testFlavorTypeUnitTest'
+            task "${testTaskName}Coverage"(type: JacocoReport, dependsOn: "$testTaskName") {
```
Comment (cthomas-newrelic) [inline jacoco.gradle:70]: Use modern task creation with `register()` method, as shown [here](https://github.com/newrelic/newrelic-android-agent/blob/develop/build.gradle#L97-L104).

**Review by cthomas-newrelic — CHANGES_REQUESTED** [review]
Looks good, but you may want to review the old GHE repo for how it used Jacoco.

File: jacoco.gradle (hunk 7 above)
Comment (cthomas-newrelic) [reply jacoco.gradle:67]: And, is that a useful log message?

File: .github/workflows/codecov.yml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        ./gradlew agent-core:jacocoTestReport
+        ./gradlew instrumentation:jacocoTestReport
+        ./gradlew plugin:gradle:jacocoTestReport
+
+    - name: Upload Test Report Part 1
+      uses: codecov/codecov-action@v3
+      with:
+        token: ${{ secrets.CODECOV_TOKEN }}
+        files: '**/testReleaseUnitTestCoverage.xml'
+
```
Comment (ywang-nr) [inline .github/workflows/codecov.yml:37]: I couldn't combine paths for the files, separated into two steps

**Comment by ywang-nr** [comment]
@cthomas-newrelic ready for review, and will squash merge. Thank you.

File: .github/workflows/codecov.yml (hunk 9 above)
Comment (cthomas-newrelic) [reply .github/workflows/codecov.yml:37]: That's fine, but the second jobs should be dependent on the first, is that correct? In that case I'd add a [needs statement](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) statement to the second job.

File: build.gradle [hunk 10]
```diff
@@ -12,6 +12,10 @@ buildscript {
         maven { url "https://plugins.gradle.org/m2/" }
         maven { url "https://oss.sonatype.org/content/repositories/comnewrelic-${project.versions.agent}" }
     }
+
+    dependencies {
+        classpath "org.jacoco:org.jacoco.core:$versions.test.jacoco"
```
Comment (cthomas-newrelic) [inline build.gradle:17]: Should wrap this as `${versions.test.jacoco}`

**Review by cthomas-newrelic — APPROVED** [review]
A few minor changes but good to go!

File: .github/workflows/codecov.yml (hunk 9 above)
Comment (ywang-nr) [reply .github/workflows/codecov.yml:37]: Actually they are all independents reports, doesn't reply on each other

**Review by cthomas-newrelic — APPROVED** [review]
Looks like you need to rebase your branch, but looks ready. Please squash commit.

**Pull request merged**